packages/common/src/cache/cache-manage-multi-request.ts

```diff
@@ -0,0 +1,136 @@
+import {Api, App} from '@yoroi/types'
+import {freeze} from 'immer'
+
+import {cacheResolveRecordsSource} from './cache-resolve-records-source'
+import {isRight} from '../utils/monads'
+import {hasEntryValue} from '../utils/predicates'
+
+export const cacheManageMultiRequest = async <K extends string, V>({
+  cachedInfosWithoutRecord,
+  ids,
+  request,
+  persistance,
+  unknownRecordFactory,
+}: {
+  cachedInfosWithoutRecord: Readonly<Map<K, App.CacheInfo | null | undefined>>
+  ids: ReadonlyArray<K>
+  request: (
+    ids: ReadonlyArray<Api.RequestWithCache<K>>,
+  ) => Promise<Readonly<Api.Response<{[key in K]: Api.ResponseWithCache<V>}>>>
+  persistance: {
+    read: (
+      ids: ReadonlyArray<K>,
+    ) => ReadonlyArray<[K, App.CacheRecord<V> | null]>
+    save: (records: ReadonlyArray<[K, App.CacheRecord<V>]>) => void
+  }
+  unknownRecordFactory?: (id: K) => Readonly<App.CacheRecord<V>>
+}) => {
+  // resolve the ids to be fetched and the ones that should be read from cache
+  const {toFetch, fromCache} = cacheResolveRecordsSource({
+    ids,
+    cachedInfosWithoutRecord,
+  })
+  let updatedIds: Array<K> = []
+
+  let toSaveRevalidatingCache: Map<K, App.CacheRecord<V>> = new Map()
+  let toRevalidateCache: Map<K, number> = new Map()
+  let toResolveLocally = new Set<K>(fromCache)
+  let cachedRecords: Map<K, App.CacheRecord<V> | null> = new Map()
+
+  let unknownIds: Array<K> = []
+  let unknownEmptyRecords: Array<[K, null]> = []
+  let toSaveNewUnknownRecord: Map<K, App.CacheRecord<V>> = new Map()
+
+  let toSaveNewFromApi: Map<K, App.CacheRecord<V>> = new Map()
+  let recordsFromApi: Record<K, Api.ResponseWithCache<V>> = {} as Record<
+    K,
+    Api.ResponseWithCache<V>
+  >
+
+  if (toFetch.length > 0) {
+    const apiResponse = await request(toFetch)
+    // if the request fails, we will show unknown tokens and wont block the user
```

Comment:
yap, that's precise, cuz it's hitting once now, next iteration is to batch it, 100/250/500 still unknown, this will mark all those tokens as unknown, but they will be stale right away, so next sync will try fetching them again, and also it can implement retry logic (at least 3 times) before building as unknown.

Reply:
like it is on tx-history right now.